// iwer/lib/views/XRView.js
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
export const PRIVATE = Symbol('@immersive-web-emulation-runtime/xr-view');
export var XREye;
(function (XREye) {
    XREye["None"] = "none";
    XREye["Left"] = "left";
    XREye["Right"] = "right";
})(XREye || (XREye = {}));
export class XRView {
    constructor(eye, projectionMatrix, transform, session) {
        this[PRIVATE] = {
            eye,
            projectionMatrix,
            transform,
            recommendedViewportScale: null,
            requestedViewportScale: 1.0,
            session,
        };
    }
    get eye() {
        return this[PRIVATE].eye;
    }
    get projectionMatrix() {
        return this[PRIVATE].projectionMatrix;
    }
    get transform() {
        return this[PRIVATE].transform;
    }
    get recommendedViewportScale() {
        return this[PRIVATE].recommendedViewportScale;
    }
    requestViewportScale(scale) {
        if (scale === null || scale <= 0 || scale > 1) {
            console.warn('Invalid scale value. Scale must be > 0 and <= 1.');
            return;
        }
        this[PRIVATE].requestedViewportScale = scale;
    }
}
