// @iwer/devui/lib/components/fov.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Button, ButtonGroup, RangeSelector } from './styled.js';
import { useState } from 'react';
import styled from 'styled-components';
const FovSettingContainer = styled.div `
	display: flex;
	justify-content: center;
	pointer-events: all;
	position: fixed;
	display: flex;
	top: 40px;
	left: calc(50vw - 156px);
	width: 312px;
`;
export const FOVMenu = ({ xrDevice, inputLayer }) => {
    const [fovy, setFovy] = useState(xrDevice.fovy);
    return (_jsx(FovSettingContainer, { children: _jsxs(ButtonGroup, { "$reverse": false, children: [_jsx(Button, { "$reverse": false, disabled: true, children: "FOV-Y" }), _jsx(RangeSelector, { "$reverse": false, value: fovy, style: { width: '100px', borderRadius: '2px' }, onChange: (e) => {
                        const value = Number(e.target.value);
                        setFovy(value);
                        xrDevice.fovy = value;
                        inputLayer.syncFovy();
                        inputLayer.renderScene();
                    }, min: Math.PI / 6, max: Math.PI / 1.5, step: Math.PI / 48 }), _jsxs(Button, { "$reverse": false, disabled: true, children: [((fovy / Math.PI) * 180).toFixed(2), "\u00B0"] })] }) }));
};
