// @iwer/devui/lib/components/header.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Button, FAIcon, MappedKeyBlock } from './styled.js';
import { faArrowsUpDown, faCaretDown, faCaretUp, faCirclePlay, faKeyboard, faRightFromBracket, faRotateLeft, faSquareArrowUpRight, faVideo, } from '@fortawesome/free-solid-svg-icons';
import { IWERIcon } from './icons.js';
import styled from 'styled-components';
const HeaderButtonsContainer = styled.div `
	padding: 6px 5px;
	display: flex;
	background-color: rgba(43, 43, 43, 0.5);
	backdrop-filter: blur(10px);
	-webkit-backdrop-filter: blur(10px);
	justify-content: center;
	pointer-events: all;
	border-radius: 0 0 12px 12px;
	align-items: center;
	height: 24px;
`;
const MovementInstructionsContainer = styled.div `
	background-color: rgba(43, 43, 43, 0.5);
	backdrop-filter: blur(10px);
	-webkit-backdrop-filter: blur(10px);
	border: none;
	display: flex;
	flex-direction: column;
	align-items: flex-start;
	justify-content: center;
	cursor: pointer;
	color: white;
	white-space: nowrap;
	font-size: 14px;
	text-transform: none;
	box-shadow: none;
	font-family: Arial, sans-serif;
	border-radius: ${({ $reverse }) => ($reverse ? '0 0 0 12px' : '0 0 12px 0')};
	padding: 5px;

	> div {
		display: flex;
		flex-direction: row;
		align-items: center;
		gap: 2px;
		margin: 2px;
	}
`;
export const HeaderUI = ({ xrDevice, inputLayer, keyMapOpen, setKeyMapOpen, fovSettingOpen, setFovSettingOpen, }) => {
    return (_jsxs("div", { style: {
            display: 'flex',
            justifyContent: 'center',
        }, children: [_jsxs(HeaderButtonsContainer, { children: [_jsx(IWERIcon, {}), _jsxs("div", { style: {
                            display: 'flex',
                            flexDirection: 'row',
                            marginLeft: '4px',
                        }, children: [_jsx(Button, { "$reverse": false, onClick: () => {
                                    inputLayer.resetDeviceTransforms();
                                }, children: _jsx(FAIcon, { icon: faRotateLeft }) }), _jsx(Button, { "$reverse": false, onClick: () => {
                                    inputLayer.lockPointer();
                                    setKeyMapOpen(false);
                                    setFovSettingOpen(false);
                                }, children: _jsx(FAIcon, { icon: faCirclePlay }) }), _jsx(Button, { "$reverse": false, onClick: () => {
                                    setKeyMapOpen(!keyMapOpen);
                                    setFovSettingOpen(false);
                                }, children: _jsx(FAIcon, { icon: faKeyboard }) }), _jsx(Button, { "$reverse": false, onClick: () => {
                                    setFovSettingOpen(!fovSettingOpen);
                                    setKeyMapOpen(false);
                                }, children: _jsx(FAIcon, { icon: faVideo }) }), _jsx(Button, { "$reverse": false, onClick: () => {
                                    const xrSession = xrDevice.activeSession;
                                    xrSession === null || xrSession === void 0 ? void 0 : xrSession.end();
                                }, children: _jsx(FAIcon, { icon: faRightFromBracket }) })] })] }), _jsxs(MovementInstructionsContainer, { "$reverse": false, style: { position: 'fixed', left: '0', top: '0' }, children: [_jsxs("div", { children: [_jsx(FAIcon, { icon: faSquareArrowUpRight, style: { marginRight: '4px' } }), ' ', "Roomscale Movement"] }), _jsxs("div", { children: [_jsx(MappedKeyBlock, { "$pressed": false, style: { width: '50px' }, children: "L Shift" }), _jsx("span", { style: { margin: '0 4px' }, children: "+" }), _jsx(MappedKeyBlock, { "$pressed": false, children: "W" }), _jsx(MappedKeyBlock, { "$pressed": false, children: "A" }), _jsx(MappedKeyBlock, { "$pressed": false, children: "S" }), _jsx(MappedKeyBlock, { "$pressed": false, children: "D" })] })] }), _jsxs(MovementInstructionsContainer, { "$reverse": true, style: { position: 'fixed', right: '0', top: '0' }, children: [_jsxs("div", { children: [_jsx(FAIcon, { icon: faArrowsUpDown, style: { marginRight: '4px' } }), " Camera Height"] }), _jsxs("div", { children: [_jsx(MappedKeyBlock, { "$pressed": false, style: { width: '50px' }, children: "L Shift" }), _jsx("span", { style: { margin: '0 4px' }, children: "+" }), _jsx(MappedKeyBlock, { "$pressed": false, children: _jsx(FAIcon, { icon: faCaretUp }) }), _jsx(MappedKeyBlock, { "$pressed": false, children: _jsx(FAIcon, { icon: faCaretDown }) })] })] })] }));
};
